Trello: Added topic for projected volumes

mburke5678 opened the pull request:
https://trello.com/c/AfFZsOyx/418-document-downward-api-all-in-one-volume-for-secrets-configmap-downward-api

mburke5678 commented:
@pmorie @jpeeler Can you take a look? 
Is this what we need from Jesse's PR (https://github.com/openshift/openshift-docs/pull/4503).

jpeeler commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,170 @@
+----
+echo -n "admin" > ./username.txt
+echo -n "1f2d1e2e67df" > ./password.txt
+----
```
Missing the secret creation commands. You could get rid of the above two lines and use the following:

secret.yaml:
```
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
type: Opaque
data:
  pass: MWYyZDFlMmU2N2Rm
  user: YWRtaW4=
---
```
oc create -f secret.yaml.

mburke5678 replied:
@jpeeler Where do the user and pass names come from?
Are they the same user/pass as in 
https://docs.openshift.com/container-platform/3.5/install_config/configuring_authentication.html

jpeeler replied:
The user and pass names are used in https://github.com/kubernetes/kubernetes.github.io/pull/3856/files#diff-92b561a119469ea9519cb2bbe83437d9R21, which I think you left out. The values are made up and can be anything valid (base64 encoded).

jpeeler commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,170 @@
+----
+oc create -f <your_yaml_file>.yaml
+----
+
```
I think the verification of examining inside the container is valuable, otherwise the "tasks" portion of this might should be excluded.

jpeeler commented:
Jesse's PR based upon my original PR is https://github.com/kubernetes/kubernetes.github.io/pull/3856/. If openshift doesn't have a "tasks" section like Kubernetes, combining them here is fine.

Made a few suggestions @mburke5678 on the commits.

mburke5678 commented:
@jpeeler I added the verification section back (just seeing the directory structure is sufficient??) and added more detail to the first example with the secrets, downward API, and configmap volume sources. And I fleshed out the use case procedure more (I used a 3.5 OpenShift system in a pinch). PTAL.

mburke5678 commented:
@chengzhang1016 Can you take a look at the document also? Trying to get this closed in

jpeeler commented on dev_guide/projected_volumes.adoc:
```diff
@@ -145,26 +145,82 @@ spec:
           name: pass
 ----
++
+The `user` and `pass` values are made up and can be anything valid (base64 encoded).
```
I don't think I'd say it quite like this. Perhaps something along the lines of, the user and pass values are for example purposes only. The examples used here are base64 encoded values "user: admin", "pass:1f2d1e2e67df".

jpeeler commented on dev_guide/projected_volumes.adoc:
```diff
@@ -145,26 +145,82 @@ spec:
+----
+
+. In your shell, verify that the `projected-volumes` directory contains your projected sources:
+
```
Ideally you'd do `ls projected-volumes` to show the projections are actually present.

mburke5678 replied:
@jpeeler What should the user see doing ls projected-volumes?

I do not have a 3.6 system at the moment (3.5 gives me an empty projected-volumes directory).

jpeeler replied:
Unfortunately after looking closer, the projections is wrong. It should be like this for the pod spec:

```
apiVersion: v1
kind: Pod
metadata:
  name: test-projected-volume
spec:
  containers:
  - name: test-projected-volume
    image: busybox
    args:
    - sleep
    - "86400"
    volumeMounts:
    - name: all-in-one
      mountPath: "/projected-volume"
      readOnly: true
  volumes:
  - name: all-in-one
    projected:
      sources:
      - secret:
          name: mysecret
```

And then you can check in the container to see:
```
/ # ls projected-volume/
pass  user
```

jpeeler commented:
I left two more comments on older commits, so click the show outdated to read.

chengzhang1016 commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,289 @@
+dev               proc              run               usr
+etc               projected-volume  sys               var
+----
+
```
@jpeeler @mburke5678 
Suggest to add some notice for customer. Such as: Collisions when one path is explicit and the other is automatically projected. You can try to explain below 2 cases:
1. Test collisions when one path is automatically projected, and then another is explicit with same path.
2. Test collisions when one path is explicit, and then another is automatically projected with same path.
There is a TC for your reference: https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-12395

mburke5678 replied:
@chengzhang1016 @jpeeler 
I took some of the collision scenarios from Jeff's document here:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/all-in-one-volume.md

Is this information appropriate for end user docs and address your request here?

Note, I also added "combinations" listed in the document's Analysis section as "Example Scenarios." Does this make sense in explaining more what projected volumes are used for?

chengzhang1016 replied:
Yes, I think the three collisions should be covered.
```
Collisions between keys when configured paths are identical.
Collisions between keys without configured paths
Collisions when one path is explicit and the other is automatically projected
```

mburke5678 commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,289 @@
+====
+The `defaultMode` can only be specified at the projected level and not for each
+volume source. However, as illustrated above, you can explicitly set the `mode`
+for each individual projection.
```
I used the configs in a TC as a basis for this change. Please let me know if this is what you expected. I added the defaultMode to both examples in this section.  
OCP-12427 - All-in-one volume: Permission mode work well in whole volume and individual resources.

chengzhang1016 replied:
That is OKay for me. Thank you.

mburke5678 commented:
@chengzhang1016 Did my changes meet your expectations? Can we move the Trello to Accepted?
https://trello.com/c/AfFZsOyx/418-document-downward-api-all-in-one-volume-for-secrets-configmap-downward-api

mburke5678 commented:
@chengzhang1016 accepted the changes via Trello.
" That is OKay for me. Thanks. Accept this card."

mburke5678 commented:
@openshift/team-documentation PTAL

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+
+== Overview
+
+A _projected volume_ maps several existing xref:volumes.adoc#dev-guide-volumes[volume sources] into the same directory.
```
Full path for the link above?

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+* automatically populate a single volume with the keys from multiple secrets, configmaps, and with downward API information,
+so that I can synthesize a single directory with various sources of information
+* populate a single volume with the keys from multiple secrets, configmaps, and with downward API information,
+explicitly specifying paths for each item, so that I can have full control over the contents of that volume
```
Should there be punctuation at the end of line 28? I'm not sure what to suggest exactly, because the sentence doesn't really run on to the two points above. Is there a way that can be clearer? I'd also suggest full stops after the points.

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+The following general scenarios show how you can use projected volumes.
+
+* *ConfigMap, Secrets, Downward API.* Projected volumes allow you to to deploy containers with configuration data that includes passwords.
+An application using these resources could be deploying OpenStack on Kubernetes. The configuration data may need to be assembled differently depending on if the services are going to be used for production or for testing. If a pod is labeled with production or testing, the downward API selector metadata.labels can be used to produce the correct OpenStack configs.
```
Markup for "metadata.labels"

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+* *Secrets + Downward API.* Projected volumes allow you to use a secret as a public key to encrypt the namespace of the pod (available via the `metadata.namespace` selector).
+This example allows the operator to use the application to deliver the namespace information securely without using an encrypted transport.
+
+## Example Pod Specifications
```
To be consistent, I'd change the ##s above (and there's a few more in the doc) to be ==s.

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+
+### Collisions between keys without configured paths
+
+The only run-time validation can occur is when all the paths are known at pod creation, similar to the above scenario.
```
Something is missing from the above. I think a "that" after "validation"?

"The only run-time validation that can occur is when all the paths are known at pod creation..."

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+for each individual projection.
+====
+
+## Pathing Considerations
```
OK, realising that the hashes are leftover from markdown now: http://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#markdown-compatibility

I'm wondering if this section would be better as a labeled list, though. It's all part of the same thing and might clog up a ToC? At the least, I'd suggest changing the #s to =s and putting caps in the titles.

mburke5678 replied:
@bfallonf Changed to labelled lists. Do we have a "proper" formatting style for labelled lists?

bfallonf replied:
Not sure if it's official, but we usually use two ::s after the heading. It come sup like this:

https://docs.openshift.com/container-platform/3.5/admin_guide/sdn_troubleshooting.html#nomenclature

That said, it's just a suggestion, and I see you've put it into dot points, which works.

mburke5678 replied:
Might I merge @bfallonf ?

bfallonf replied:
@mburke5678  Sure, merge away.

bfallonf commented on dev_guide/projected_volumes.adoc:
```diff
@@ -0,0 +1,368 @@
+
+The following example shows how to use a projected volume to mount an existing Secret volume source.
+
+The steps can be used to create a user name and password xref:secrets.adoc#dev-guide-secret[Secrets] from local files. You then create a pod that runs one container, using a projected volume to mount the Secrets into the same shared directory.
```
Full path for the link above?

bfallonf commented:
@mburke5678 I made some comments. Mainly formatting stuff. Let me know what you think.

mburke5678 commented:
[rev_history]
|xref:../dev_guide/projected_volumes.adoc#dev-guide-projected-volumes[Developer Guide -> Projected Volumes]
|Added new section on xref:../dev_guide/projected_volumes.adoc#dev-guide-projected-volumes[Kubernetes projected volumes] to the Developer Guide.
%

adellape commented:
@mburke5678 The section titles in this could use [unique IDs/anchors](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#unique-ids).

mburke5678 commented:
See follow-up for IDs https://github.com/openshift/openshift-docs/pull/4601
